src/arm/midr.c

```diff
@@ -21,6 +25,66 @@
 #include "uarch.h"
 #include "sve.h"
 
+
+#if defined _WIN32
+// Windows stores processor information in registery at:
+// "HKEY_LOCAL_MACHINE\HARDWARE\DESCRIPTION\System\CentralProcessor"
+// Within this directory, there each core will get its own folder as well with
+// registery entries that map to system registers on ARM in the `CP ####` registry entries
+// Ex. the MIDR register for core 0 is the `REG_QWORD` at:
+// "HKEY_LOCAL_MACHINE\HARDWARE\DESCRIPTION\System\CentralProcessor\0\CP 4000"
+// The name of these `CP ####`-registers follow their register ID encoding in hexadecimal
+// (op0&1):op1:crn:crm:op2.
+// More registers can be found here:
+// https://developer.arm.com/documentation/ddi0601/2024-06/AArch64-Registers
+// Some important ones:
+// CP 4000: MIDR_EL1
+// CP 4020: ID_AA64PFR0_EL1
+// CP 4021: ID_AA64PFR1_EL1
+// CP 4028: ID_AA64DFR0_EL1
+// CP 4029: ID_AA64DFR1_EL1
+// CP 402C: ID_AA64AFR0_EL1
+// CP 402D: ID_AA64AFR1_EL1
+// CP 4030: ID_AA64ISAR0_EL1
+// CP 4031: ID_AA64ISAR1_EL1
+// CP 4038: ID_AA64MMFR0_EL1
+// CP 4039: ID_AA64MMFR1_EL1
+// CP 403A: ID_AA64MMFR2_EL1
+
+bool read_registry_hklm_32(char* subkey, char* name, int* value) {
```

Comment:
Why there are _32 and _64 variants of `read_registry_hklm` and `get_win32_core_info`?

Reply:
They particularly differ due to how the registry-entries themselves are different types(`RRF_RT_REG_DWORD` vs. `RRF_RT_REG_QWORD`) and will write to `value` differently(4 bytes vs 8 bytes).

Reply:
In such case, could we do a refactoring like this? From:


```c
bool read_registry_hklm_64(char* subkey, char* name, int64_t* value) {
  DWORD value_len = sizeof(int64_t);
  if(RegGetValueA(HKEY_LOCAL_MACHINE, subkey, name, RRF_RT_REG_QWORD, NULL, value, &value_len) != ERROR_SUCCESS) {
    printBug("Error reading registry entry \"%s\\%s\"", subkey, name);
    return false;
  }
  return true;
}

bool read_registry_hklm_32(char* subkey, char* name, int* value) {
  DWORD value_len = sizeof(int);
  if(RegGetValueA(HKEY_LOCAL_MACHINE, subkey, name, RRF_RT_REG_DWORD, NULL, value, &value_len) != ERROR_SUCCESS) {
    printBug("Error reading registry entry \"%s\\%s\"", subkey, name);
    return false;
  }
  return true;
}
```

to

```c
bool read_registry_hklm(char* subkey, char* name, void* value, bool is64) {  
  DWORD value_len;
  int foo;

  if (is64) {
   value_len = sizeof(int64_t);
   foo = RRF_RT_REG_QWORD;
  }
  else {
    value_len = sizeof(int);
    foo = RRF_RT_REG_DWORD;
  }

  if(RegGetValueA(HKEY_LOCAL_MACHINE, subkey, name, foo, NULL, value, &value_len) != ERROR_SUCCESS) {
    printBug("Error reading registry entry \"%s\\%s\"", subkey, name);
    return false;
  }
  return true;
}

bool read_registry_hklm_64(char* subkey, char* name, int64_t* value) {
  return read_registry_hklm(subkey, name, value, true);
}


bool read_registry_hklm_32(char* subkey, char* name, int* value) {
  return read_registry_hklm(subkey, name, value, false);
}
```

(untested, I have no idea about Windows APIs so feel free to fix or just reject it)

Reply:
Ended up refactoring it like this to avoid code duplication:
```c

bool read_registry_hklm_int(char* path, char* name, void* value, bool is64) {  
  DWORD value_len;
  int reg_type;
  if (is64) {
	value_len = sizeof(int64_t);
	reg_type = RRF_RT_REG_QWORD;
  }
  else {
	value_len = sizeof(int32_t);
	reg_type = RRF_RT_REG_DWORD;
  }

  if(RegGetValueA(HKEY_LOCAL_MACHINE, path, name, reg_type, NULL, value, &value_len) != ERROR_SUCCESS) {
    printBug("Error reading registry entry \"%s\\%s\"", path, name);
    return false;
  }
  return true;
}

bool get_win32_core_info_int(uint32_t core_index, char* name, void* value, bool is64) {
  // path + digits
  uint32_t max_path_size = 45+3+1;
  char* path = emalloc(sizeof(char) * max_path_size);
  snprintf(path, max_path_size, "HARDWARE\\DESCRIPTION\\System\\CentralProcessor\\%u", core_index);
  return read_registry_hklm_int(path, name, value, is64);
}
```

Reply:
Excellent!